implement oog constant state (#266)
privacy-scaling-explorations/zkevm-specs: specs/error_state/ErrorOutOfGasConstant.md

# ErrorOutOfGasConstant state

## Procedure
For opcodes which have non-zero constant gas cost, this type of error occurs when gas left of 
current step is less than required constant gas .

### EVM behavior

1. If it's a root call, it ends the execution.
2. Otherwise, it restores caller's context and switch to it.

> ### Circuit behavior
> 
> The circuit checks following items:
> 
> 1. remaining gas left is indeed less than this step's constant gas required.
> 2. current call must be failed.
> 3. If it's a root call, it transits to `EndTx`, and the call's `IsPersistent` must be false
> 4. if it is not root call, it restores caller's context by reading to `rw_table`, then does step state transition to it.

Reviewer: I would structure this a bit different to make it more clear:
```suggestion
### Constraints
1. `remaining_gas_left < step_constant_gas_required`.
2. Current call must be failed.
3. If it's a root call, it transits to `EndTx`, and the call's `IsPersistent` must be false
4. if it is not root call, it restores caller's context by reading to `rw_table`, then does step state transition to it.

### Lookups
- 1 Call Context lookup `CallContextFieldTag.IsSuccess`.
- 1 Call Context lookup `CallContextFieldTag.IsPersistent` (Only if Current instruction is root.). 
```
